hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -777,29 +787,60 @@ int SaraNcpClient::waitAtResponse(unsigned int timeout, unsigned int period) {
     return SYSTEM_ERROR_TIMEOUT;
 }
 
-int SaraNcpClient::waitReady() {
+int SaraNcpClient::waitReady(bool powerOn) {
     if (ready_) {
         return 0;
     }
-    muxer_.stop();
-    CHECK(serial_->setBaudRate(UBLOX_NCP_DEFAULT_SERIAL_BAUDRATE));
-    CHECK(initParser(serial_.get()));
-    // Enable voltage translator
+
+    ModemState modemState = ModemState::Unknown;
+
+    // Just in case make sure that the voltage translator is on
+    CHECK(modemSetUartState(true));
+    HAL_Delay_Milliseconds(100);
+    CHECK(modemSetUartState(false));
+    HAL_Delay_Milliseconds(100);
     CHECK(modemSetUartState(true));
-    skipAll(serial_.get(), 1000);
-    parser_.reset();
-    ready_ = waitAtResponse(20000) == 0;
+
+    if (powerOn) {
+        LOG_DEBUG(TRACE, "Waiting for modem to be ready from power on");
+        CHECK(defaultParserConfig());
+        ready_ = waitAtResponse(20000) == 0;
+        modemState = ModemState::DefaultBaudrate;
+    } else if (ncpState() == NcpState::OFF) {
+        LOG_DEBUG(TRACE, "Waiting for modem to be ready from current unknown state");
+        ready_ = checkRuntimeState(modemState) == 0;
+        if (ready_) {
+            LOG_DEBUG(TRACE, "Runtime state %d", (int)modemState);
+        }
+    } else {
+        // Most likely we just had a parser error, try to get a response from the modem as-is
+        LOG_DEBUG(TRACE, "Waiting for modem to be ready after parser error");
```

Comment:
Would like to replace the comment with "likely a parser error". Might be easy for debugging in future, if anything?